lib/rexml/parsers/baseparser.rb

```diff
@@ -614,87 +613,60 @@ def process_instruction(start_position)
-            end
-            unless scanner.scan(/.*#{Regexp.escape(quote)}/um)
-              @source.ensure_buffer
-              match_data = @source.match(/^(.*?)(\/)?>/um, true)
-              if match_data
-                scanner << "/" if closed
-                scanner << ">"
-                scanner << match_data[1]
-                scanner.pos = start_position
-                closed = !match_data[2].nil?
-                next
+            unless match = @source.match(/(['"])(.*?)\1\s*/um, true)
+              if match = @source.match(/(['"])/, true)
+                message =
+                  "Missing attribute value end quote: <#{name}>: <#{match[1]}>"
+                raise REXML::ParseException.new(message, @source)
+              else
+                message = "Missing attribute value start quote: <#{name}>"
+                raise REXML::ParseException.new(message, @source)
               end
-              message =
-                "Missing attribute value end quote: <#{name}>: <#{quote}>"
-              raise REXML::ParseException.new(message, @source)
             end
-          end
-          name = scanner[1]
-          prefix = scanner[2]
-          local_part = scanner[3]
-          # quote = scanner[4]
-          value = scanner[5]
-          if prefix == "xmlns"
-            if local_part == "xml"
-              if value != "http://www.w3.org/XML/1998/namespace"
-                msg = "The 'xml' prefix must not be bound to any other namespace "+
+            value = match[2]
+            if prefix == "xmlns"
+              if local_part == "xml"
+                if value != "http://www.w3.org/XML/1998/namespace"
+                  msg = "The 'xml' prefix must not be bound to any other namespace "+
+                    "(http://www.w3.org/TR/REC-xml-names/#ns-decl)"
+                  raise REXML::ParseException.new( msg, @source, self )
+                end
+              elsif local_part == "xmlns"
+                msg = "The 'xmlns' prefix must not be declared "+
                   "(http://www.w3.org/TR/REC-xml-names/#ns-decl)"
-                raise REXML::ParseException.new( msg, @source, self )
+                raise REXML::ParseException.new( msg, @source, self)
               end
-            elsif local_part == "xmlns"
-              msg = "The 'xmlns' prefix must not be declared "+
-                "(http://www.w3.org/TR/REC-xml-names/#ns-decl)"
-              raise REXML::ParseException.new( msg, @source, self)
+              curr_ns << local_part
+            elsif prefix
+              prefixes << prefix unless prefix == "xml"
             end
-            curr_ns << local_part
-          elsif prefix
-            prefixes << prefix unless prefix == "xml"
-          end
 
-          if attributes.has_key?(name)
-            msg = "Duplicate attribute #{name.inspect}"
-            raise REXML::ParseException.new(msg, @source, self)
-          end
+            if attributes[name]
```

Comment:
Why did you change `.has_key?` to `[]`?

Reply:
I changed it to `attributes[name]` rather than `attributes.has_key?(name)` because it is faster to process and shorter to write.
Should I not have mixed it with this PR?

```ruby
$ cat benchmark/attributes.yaml 
loop_count: 100000
contexts:
  - name: No YJIT
    prelude: |
      $LOAD_PATH.unshift(File.expand_path("lib"))
      require 'rexml'
  - name: YJIT
    prelude: |
      $LOAD_PATH.unshift(File.expand_path("lib"))
      require 'rexml'
      RubyVM::YJIT.enable

prelude: |
  attributes = {}
  name = :a

benchmark:
  'attributes[name]' : attributes[name]
  'attributes.has_key?(name)' : attributes.has_key?(name)
```

```bash
$ benchmark-driver benchmark/attributes.yaml 
Calculating -------------------------------------
                             No YJIT        YJIT 
         attributes[name]    51.680M     54.437M i/s -    100.000k times in 0.001935s 0.001837s
attributes.has_key?(name)    46.019M     45.683M i/s -    100.000k times in 0.002173s 0.002189s

Comparison:
                      attributes[name]
                     YJIT:  54436580.8 i/s 
                  No YJIT:  51679591.0 i/s - 1.05x  slower

             attributes.has_key?(name)
                  No YJIT:  46019327.8 i/s 
                     YJIT:  45682957.4 i/s - 1.01x  slower
```

Reply:
Oh... It's surprising...

Both of them just calls `hash_stlike_lookup()`:

`Hash#[]`: https://github.com/ruby/ruby/blob/b4f3f3c1031cc9ef5c6741042236db497be6602b/hash.c#L2135-L2146
`Hash#has_key?`: https://github.com/ruby/ruby/blob/b4f3f3c1031cc9ef5c6741042236db497be6602b/hash.c#L3671-L3675

Could you split this to another PR for now? I want to take a look at it later.

Reply:
OK. I See.